File: _23_Convert_Sorted_Array_to_Binary_Search_Tree.java
public class _23_Convert_Sorted_Array_to_Binary_Search_Tree {
    public static void main(String[] args) {
        Solution sol = new Solution();

        int[] nums1 = {-10, -3, 0, 5, 9};
        TreeNode root1 = sol.sortedArrayToBST(nums1);
        System.out.print("Ejemplo 1 ([-10, -3, 0, 5, 9]): ");
        printInOrder(root1); 
        System.out.println();

        int[] nums2 = {1, 3};
        TreeNode root2 = sol.sortedArrayToBST(nums2);
        System.out.print("Ejemplo 2 ([1, 3]): ");
        printInOrder(root2); 
        System.out.println();

        int[] nums3 = {0};
        TreeNode root3 = sol.sortedArrayToBST(nums3);
        System.out.print("Ejemplo 3 ([0]): ");
        printInOrder(root3); 
        System.out.println();
    }

    // Función de apoyo: imprime el árbol en orden (inorder traversal)
    public static void printInOrder(TreeNode root) {
        if (root == null) return;
        printInOrder(root.left);
        System.out.print(root.val + " ");
        printInOrder(root.right);
    }
}


class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}


class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return buildTree(nums, 0, nums.length - 1);
    }

    private TreeNode buildTree(int[] nums, int left, int right) {
        if (left > right) return null; // caso base

        int mid = left + (right - left) / 2; // para evitar overflow
        TreeNode root = new TreeNode(nums[mid]);

        root.left = buildTree(nums, left, mid - 1);   // construir izquierda
        root.right = buildTree(nums, mid + 1, right); // construir derecha

        return root;
    }
}

File: _23_Convert_Sorted_Array_to_Binary_Search_Tree.ts

class TreeNode5 {
    val: number;
    left: TreeNode5 | null;
    right: TreeNode5 | null;

    constructor(val?: number, left?: TreeNode5 | null, right?: TreeNode5 | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

function sortedArrayToBST(nums: number[]): TreeNode5 | null {
    if (nums.length === 0) return null;

    function buildBST(left: number, right: number): TreeNode5 | null {
        if (left > right) return null;

        const mid = Math.floor((left + right) / 2);
        const root = new TreeNode5(nums[mid]);

        root.left = buildBST(left, mid - 1);
        root.right = buildBST(mid + 1, right);

        return root;
    }

    return buildBST(0, nums.length - 1);
}


function inorderTraversal(root: TreeNode5 | null, arr: number[]): void {
    if (!root) return;
    inorderTraversal(root.left, arr);
    arr.push(root.val);
    inorderTraversal(root.right, arr);
}


function main() {
    const nums1 = [-10, -3, 0, 5, 9];
    const root1 = sortedArrayToBST(nums1);
    const result1: number[] = [];
    inorderTraversal(root1, result1);
    console.log("Ejemplo 1 (inorder):", result1.join(" "));

    const nums2 = [1, 3];
    const root2 = sortedArrayToBST(nums2);
    const result2: number[] = [];
    inorderTraversal(root2, result2);
    console.log("Ejemplo 2 (inorder):", result2.join(" "));
}

main();
